tests/pootle_fs/fs_response.py

```diff
@@ -0,0 +1,206 @@
+        if "fs_path" in self.kwargs:
+            return self.kwargs["fs_path"]
+        elif self.store_fs:
+            return self.store_fs.path
+
+    @property
+    def store(self):
+        if "store" in self.kwargs:
+            return self.kwargs["store"]
+        elif self.store_fs:
+            return self.store_fs.store
+
+    @property
+    def store_fs(self):
+        return self.kwargs.get("store_fs")
+
+
+class DummyFSState(State):
+    """The pootle_fs State can create ItemStates with
+    - a store_fs (that has a store)
+    - a store_fs (that has no store)
+    - a store and an fs_path
+    - a pootle_path and an fs_path
+    """
+
+    item_state_class = DummyFSItemState
+
+    def state_fs_added(self, **kwargs):
+        for store_fs in kwargs.get("fs_added", []):
+            yield dict(store_fs=store_fs)
+
+    def state_fs_ahead(self, **kwargs):
+        for store_fs in kwargs.get("fs_ahead", []):
+            yield dict(store_fs=store_fs)
+
+    def state_fs_untracked(self, **kwargs):
+        for fs_path, pootle_path in kwargs.get("fs_untracked", []):
+            yield dict(fs_path=fs_path, pootle_path=pootle_path)
+
+    def state_pootle_untracked(self, **kwargs):
+        for fs_path, store in kwargs.get("pootle_untracked", []):
+            yield dict(fs_path=fs_path, store=store)
+
+
+@pytest.mark.django_db
+def test_fs_response_instance():
+    context = DummyContext()
+    resp = ProjectFSResponse(context)
+    assert resp.context == context
+    assert resp.response_types == FS_RESPONSE.keys()
+    assert resp.has_failed is False
+    assert resp.made_changes is False
+    assert list(resp.failed()) == []
+    assert list(resp.completed()) == []
+    assert str(resp) == (
+        "<ProjectFSResponse(<DummyContext object>): No changes made>")
+    assert list(resp) == []
+    with pytest.raises(KeyError):
+        resp["DOES_NOT_EXIST"]
+
+
+def _test_item(item, item_state):
+    assert isinstance(item, ProjectFSItemResponse)
+    assert item.kwargs["fs_state"] == item_state
+    assert item.fs_state == item_state
+    assert item.failed is False
+    assert item.fs_path == item.fs_state.fs_path
+    assert item.pootle_path == item.fs_state.pootle_path
+    assert item.store_fs == item.fs_state.store_fs
+    assert item.store == item.fs_state.store
+    assert (
+        str(item)
+        == ("<ProjectFSItemResponse(<DummyContext object>): %s "
+            "%s::%s>" % (item.action_type, item.pootle_path, item.fs_path)))
+
+
+def _test_fs_response(**kwargs):
+    action_type = kwargs.pop("action_type")
+    state_type = kwargs.pop("state_type")
+    resp = ProjectFSResponse(DummyContext())
+    state = DummyFSState(DummyContext(), **kwargs)
+    for fs_state in state[state_type]:
+        resp.add(action_type, fs_state=fs_state)
+    assert resp.has_failed is False
+    assert resp.made_changes is True
+    assert resp.response_types == FS_RESPONSE.keys()
+    assert len(list(resp.completed())) == 2
```

Comment:
ive added it as a param to the test function, but i dont think i need to pass it every time